### ISSUE TITLE
PanZoom does not work with pandas datetime data - x-axis overzooms

### ISSUE THREAD

**Issue by acelere** [bqplot/bqplot#617]
Trying to panzoom a linechart generated with the attached csv file.
When using the zoom function, the x-axis (DateScale) goes completely off, making this feature unusable. For one click in the wheel, the scale changes by more than 5 hours. I was expecting changes in the order of minutes.
Attached is a reference notebook implementation showing the problem.
[bqplot_panzoom_issue.zip](https://github.com/bloomberg/bqplot/files/1753008/bqplot_panzoom_issue.zip)
You will need to change your file path to import the data, currently set to: 
`file_path` = `'~/work/FCMC_10Hz_simple.csv'`

This is the data file:
[FCMC_10Hz_simple.zip](https://github.com/bloomberg/bqplot/files/1753010/FCMC_10Hz_simple.zip)

The last cell is my current, inelegant, workaround, where I need to plot my data against numeric index instead of its true time stamp.

Could you look into this? I could not find what the problem is. I tried to set the min= during my DateScale declaration, but that did not help either.

Help is much appreciated.

Tks.

**Comment by acelere** [comment]
@DougRzz , I tried your code (verbatim) on Chrome, Firefox and Edge. They all give me the same result:
![zoom_gif_w10_chrome](https://user-images.githubusercontent.com/12059585/36872286-c474db44-1d72-11e8-9ef4-ac578b95bbed.gif)
Notice how my time axis changes from 10:00 to 03:00 on the first click.
Do you think this maybe a timezone problem? Or a computer setting related to timezone?
I run my notebook on a jupyterhub server (Ubuntu based) which spawns a docker container (also Ubuntu based). But in the past, I ran also the notebook server locally and had the exact same problem.

Let me know how I can help troubleshooting.
bqplot is AWESOME and I really think it is a great idea. I want to help improve it.

Here is my `!pip freeze` contents:
alembic==0.9.8
bleach==2.1.2
bqplot==0.10.5
certifi==2018.1.18
chardet==3.0.4
cycler==0.10.0
decorator==4.2.1
entrypoints==0.2.3
html5lib==1.0.1
idna==2.6
ipykernel==4.8.2
ipython==6.2.1
ipython-genutils==0.2.0
ipywidgets==7.1.2
jedi==0.11.1
Jinja2==2.10
jsonschema==2.6.0
jupyter-client==5.2.2
jupyter-core==4.4.0
jupyterhub==0.8.1
Mako==1.0.7
MarkupSafe==1.0
matplotlib==2.1.2
mistune==0.8.3
nbconvert==5.3.1
nbformat==4.4.0
notebook==5.4.0
numpy==1.14.1
pamela==0.3.0
pandas==0.22.0
pandocfilters==1.4.2
parso==0.1.1
pexpect==4.4.0
pickleshare==0.7.4
prompt-toolkit==1.0.15
ptyprocess==0.5.2
Pygments==2.2.0
pyparsing==2.2.0
python-dateutil==2.6.1
python-editor==1.0.3
python-oauth2==1.1.0
pytz==2018.3
pyzmq==17.0.0
requests==2.18.4
scipy==1.0.0
Send2Trash==1.5.0
simplegeneric==0.8.1
six==1.11.0
SQLAlchemy==1.2.3
terminado==0.8.1
testpath==0.3.1
tornado==4.5.3
traitlets==4.3.2
traittypes==0.0.6
urllib3==1.22
wcwidth==0.1.7
webencodings==0.5.1
widgetsnbextension==3.1.4


Tks

**Comment by DougRzz** [comment]
Its a bug which I can't replicate.  I'm not sure what to suggest.  Are you able to try it on Mac or Windows?  Maybe show that animation on the BQplot Gitter channel and one of the developers might be able to help.

**Comment by SylvainCorlay** [comment]
Thanks for looking into this @DougRzz

**Comment by acelere** [comment]
Hello folks, I just installed jupyter locally on a windows machine and I get the exactly same bad result I had before. Chrome or Edge or Firefox.
Like I said, what seems really weird to me is that the scaling seems to change the minutes as expected, however it adds 5 hours to it... which smells like timezone issue. But I don't know what to do/try.
I printed the data min value and the scale min values. See below.
Running copy paste code from @DougRzz


Tks.


![zoom_gif_native_w10_chrome](https://user-images.githubusercontent.com/12059585/36900532-a055b510-1df1-11e8-9a40-f6cdfc146d13.gif)

**Comment by acelere** [comment]
Follow-up.
OK, so I went into Windows>Settings>Date&Time 
Under the "Time Zone" dropdown, I had "(UTC -05:00) Eastern Time (US & Canada)" 
Bingo, when I change it to " (UTC) Coordinated Universal Time", zooming works as posted above.
@DougRzz , please try to reproduce this.
Let me know...

**Comment by DougRzz** [comment]
Yep.  When i change the timezone, it messes up the zoom.   So zooming will only work for those living in Iceland, UK, and Portugal.  A bit unfair I say.

**Comment by acelere** [comment]
ok. should I raise a bug somewhere else?
tks for helping me out!

**Comment by DougRzz** [comment]
@acelere just leave it here. I might soon be demonstrating my bqplot dashboard with zoomable date scale to my colleagues in the US. So thanks for spotting that one.

**Comment by DougRzz** [comment]
My Javascript knowledge is zero.  But i had a look at the Panzoom class and it led me to BaseModel.js.  I would guess that the problem lies in this bit of code.......

```javascript
convert_to_json: function(elem) {
	// converts the date to a json compliant format
	if(elem === undefined || elem === null) {
		return null;
	} else {
		if (elem.toJSON === undefined) {
			return elem;
		} else {
			// the format of the string to be sent across is
			// '%Y-%m-%dT%H:%M:%S.%f'
			// by default, toJSON returns '%Y-%m-%dT%H:%M:%S.%uZ'
			// %u is milliseconds. Hence adding 000 to convert it into
			// microseconds.

			return elem.toJSON().slice(0, -1) + '000';
		}
	}
}
```

So Json by default returns '%Y-%m-%dT%H:%M:%S.%uZ' where the Z indicates it has been converted to UTC.  The Z is removed and then the  milli seconds is converted to micro seconds, but there is nothing to now indicate it is a time relative to UTC.   Its just a guess.    I maybe on completely the wrong track.  Does anyone have any suggestions?

**Comment by acelere** [comment]
Yes, I agree.
I wonder why they need to set it to a JSON compliant date.
Still in BaseModel.js:

```
set_date_elem: function(param, value) {

    this.set(param, this.convert_to_json(value));

},
``` 
Perhaps try something like:  
```
set_date_elem: function(param, value) {

    this.set(param, value);

},
``` 
and see what happens...
But I have also crude understanding of JavaScript...

Tks.

**Comment by acelere** [comment]
It appears similar issues have been around since 2015: [this old thead](https://github.com/ipython/ipykernel/pull/18) where jasongrout said:

> By the way, IIRC, there is a difference between how python and how browsers handle local vs. UTC. If I remember correctly, we were having issues where python converted things to UTC, and the browser assumed it was local time, or vice versa, if the timezone offset wasn't explicitly specified. There's a whole can of worms there.

**Comment by acelere** [comment]
I did some further investigation and it is not just the IndexSelector and Panzoom that have this issue. If I modify the ["Selectors.ipynb"](https://github.com/bloomberg/bqplot/blob/master/examples/Interactions/Selectors.ipynb) example in the repo by setting the index of the dataframe to something that is a Timestamp, the line graph shows me the wrong time (x-axis):

```
#original code
#price_data = pd.DataFrame(np.cumsum(np.random.randn(150, 2).dot([[0.5, 0.4], [0.4, 1.0]]), axis=0) + #100,
#                          columns=[symbol, symbol2],
#                          index=pd.date_range(start='11-01-2019', periods=150))

#modified code
start = pd.Timestamp('2019-11-01T16:00:00')
end = pd.Timestamp('2019-11-01T17:00:00')
t = np.linspace(start.value, end.value, 150)
t = pd.to_datetime(t)
price_data = pd.DataFrame(np.cumsum(np.random.randn(150, 2).dot([[0.5, 0.4], [0.4, 1.0]]), axis=0) + 100,
                          columns=[symbol, symbol2],
                          index=t)
````

Notice that my data starts at 2019-11-11 16:00:00
My timezone (on windows OS) is set to UTC-5

and I get this graph:

![bqplot_date_issue](https://user-images.githubusercontent.com/12059585/68489941-18fd2200-0216-11ea-8390-89f94bb1f62b.PNG)

I tried changing the timezone on the OS and it does affect the date that I get on the graph.

I am running bqplot version 0.12.0

**Comment by joders** [comment]
Another minimal replicating example:

```
import importlib
import bqplot
importlib.reload(bqplot)
from bqplot import *
xsc=DateScale()
ysc=LinearScale()
pz=PanZoom(scales={'x':[xsc],'y':[ysc]})
mark=Scatter(scales={'x':xsc,'y':ysc})
xax=Axis(scale=xsc)
yax=Axis(scale=ysc,orientation='vertical')
fig = Figure(marks=[mark],axes=(xax,yax),interaction=pz)
mark.x=np.array([
    "2019-01-01T00:00:00",
    "2019-01-02T00:00:00",
    "2019-01-03T00:00:00",
]).astype("datetime64[s]")
mark.y=[0,1,0]
fig
```

**Comment by joders** [comment]
I also tested the fix with the initial example given above

**Comment by acelere** [comment]
This issue has been fixed by **[PR#968](https://github.com/bqplot/bqplot/pull/968)** merged by commit afbe2db into masters. (bqplot version 0.12.8). Thanks for the fix @joders !
